ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFDateDiff.java

```diff
@@ -64,121 +66,46 @@
         + "  1")
 @VectorizedExpressions({VectorUDFDateDiffColScalar.class, VectorUDFDateDiffColCol.class, VectorUDFDateDiffScalarCol.class})
 public class GenericUDFDateDiff extends GenericUDF {
-  private transient Converter inputConverter1;
-  private transient Converter inputConverter2;
+  private final transient Converter[] tsConverters = new Converter[2];
   private IntWritable output = new IntWritable();
-  private transient PrimitiveCategory inputType1;
-  private transient PrimitiveCategory inputType2;
-  private IntWritable result = new IntWritable();
+  private final transient PrimitiveCategory[] tsInputTypes = new PrimitiveCategory[2];
+
 
   public GenericUDFDateDiff() {
   }
 
   @Override
   public ObjectInspector initialize(ObjectInspector[] arguments) throws UDFArgumentException {
-    if (arguments.length != 2) {
-      throw new UDFArgumentLengthException(
-        "datediff() requires 2 argument, got " + arguments.length);
-    }
-    inputConverter1 = checkArguments(arguments, 0);
-    inputConverter2 = checkArguments(arguments, 1);
-    inputType1 = ((PrimitiveObjectInspector) arguments[0]).getPrimitiveCategory();
-    inputType2 = ((PrimitiveObjectInspector) arguments[1]).getPrimitiveCategory();
-    ObjectInspector outputOI = PrimitiveObjectInspectorFactory.writableIntObjectInspector;
-    return outputOI;
+    checkArgsSize(arguments,2,2);
+    checkArgPrimitive(arguments, 0);
+    checkArgPrimitive(arguments, 1);
+    checkArgGroups(arguments, 0, tsInputTypes, STRING_GROUP, DATE_GROUP);
+    checkArgGroups(arguments, 1, tsInputTypes, STRING_GROUP, DATE_GROUP);
+    obtainTimestampConverter(arguments,0, tsInputTypes, tsConverters);
+    obtainTimestampConverter(arguments,1, tsInputTypes, tsConverters);
```

Comment:
Does it make sense to refactor as below?
```
for (int i = 0; i < arguments.length; i++) {
  checkArgPrimitive(...)
  chechArgGroups(...)
  obtainTimestampConverter(...)
}
```

Reply:
Done